crates/context/config/src/client/env/proxy/types/starknet.rs

```diff
@@ -397,3 +446,143 @@ impl From<StarknetProposalId> for ProposalId {
+                        }
+                    }
+                    entries.push((key, value));
+                }
+            }
+        }
+
+        Ok(Self { entries })
+    }
+}
+
+impl From<(Vec<Felt>, Vec<Felt>)> for ContextStorageEntry {
+    fn from((key_felts, value_felts): (Vec<Felt>, Vec<Felt>)) -> Self {
+        let key = key_felts
+            .iter()
+            .flat_map(|f| f.to_bytes_be())
+            .filter(|&b| b != 0)
+            .collect();
+
+        let value = value_felts
+            .iter()
+            .flat_map(|f| f.to_bytes_be())
+            .filter(|&b| b != 0)
+            .collect();
+
+        ContextStorageEntry { key, value }
+    }
+}
+
+struct ValueCodec<'a>(&'a serde_json::Value);
+
+impl<'a> Encode for ValueCodec<'a> {
+    fn encode<W: FeltWriter>(&self, writer: &mut W) -> Result<(), Error> {
+        match self.0 {
+            serde_json::Value::Bool(b) => {
+                writer.write(Felt::from(*b as u64));
+                Ok(())
+            }
+            serde_json::Value::String(s) => {
+                if s.starts_with("0x") {
+                    // Handle hex string directly as a single Felt
+                    writer.write(
+                        Felt::from_hex(&s)
+                            .map_err(|e| Error::custom(&format!("Invalid hex string: {}", e)))?,
+                    );
+                } else {
+                    // Regular string - split into chunks
+                    let chunk_size = 31;
+                    let chunks: Vec<_> = s.as_bytes().chunks(chunk_size).collect();
+
+                    // Write number of chunks first
+                    writer.write(Felt::from(chunks.len()));
+
+                    // Write each chunk as a Felt
+                    for chunk in chunks {
+                        writer.write(Felt::from_bytes_be_slice(chunk));
+                    }
+                }
+                Ok(())
+            }
+            serde_json::Value::Array(arr) => {
+                // Write array length first
+                writer.write(Felt::from(arr.len()));
+
+                // Encode each array element
+                for item in arr {
+                    ValueCodec(item).encode(writer)?;
+                }
+                Ok(())
+            }
+            serde_json::Value::Number(n) => {
+                if let Some(n) = n.as_u64() {
+                    writer.write(Felt::from(n));
+                } else {
+                    // For other numbers, convert to string and encode as string
+                    let s = n.to_string();
+                    ValueCodec(&serde_json::Value::String(s)).encode(writer)?;
+                }
+                Ok(())
+            }
+            serde_json::Value::Object(obj) => {
+                writer.write(Felt::from(obj.len()));
+                for (key, value) in obj {
+                    ValueCodec(&serde_json::Value::String(key.clone())).encode(writer)?;
```

Comment:
to avoid the clone here, you can move the logic for the `String` case into a function which you can call here.

signature can be something like `fn(&str, &mut Vec<Felt>)`, since string encoding is (or rather should be), infallible